### ISSUE TITLE
DIE for the "this" parameter is missing `DW_AT_location` with the new PM under `-Og`

### ISSUE THREAD

**Issue by petrhosek** [llvm/llvm-project#54290]
With the following test case:

```
extern "C" int printf(const char*,...);

class Printer {
public:
Printer(int start, int step, int end) : start_(start), step_(step), end_(end)
{}
void Run() {
for (int i = start_; i < end_; i += step_)
printf("%d\n", i); // line 8
}

private:
int start_;
int step_;
int end_;
};

int main() {
Printer p(1, 2, 10);
p.Run();
}
```

DIE for the "this" parameter is missing `DW_AT_location` when compiled with `-Og`. This used to work properly prior to Clang 13.0.0. We ran a bisect and found out this issue was introduced by the new PM.

The reason this showed up in Clang 13.0.0 is because that's when [new PM became the default](https://reviews.llvm.org/D95380), but the issue reproduces all the way back to https://github.com/llvm/llvm-project/commit/50f9e893f2c5c5c820da8d552ca8a64cd0029d65 it you pass `-fexperimental-new-pass-manager`.

**Comment by jmorse** [comment]
Looks like everything gets inlined into the `main` function, after which the `p` object gets SROA'd and all its fields optimised away. Using clang-13 with `-fno-experimental-new-pass-manager`, the major difference seems to be that fewer things are inlined, thus `p` remains allocated on the stack. Thus the major part of this variable going missing is that more optimisation is being applied, and it's now completely eliminated.

This might be the kind of thing that DWARFs "implicit pointer" facility might be able to describe, i.e. describing the values a pointer would have pointed at, despite it no longer existing in the program. There was an implementation of this a couple of years ago in the chain of patches at D84120, however it stalled alas.

**Comment by frobtech** [comment]
This is exactly the case that DW_OP_implicit_pointer is meant to handle.

**Comment by liudangyi** [comment]
Is it fully solvable by `DW_OP_implicit_pointer`? Or should we just disable this optimization in `-Og`? To be more clear the requirement is the debugger be able to print the value of `start_` at line 8.

**Comment by frobtech** [comment]
Correct DWARF for this case using DW_OP_implicit_pointer provides complete information for a debugger that implements the full DWARF semantics.

**Comment by EricRahm** [comment]
@frobtech @jmorse [`DW_OP_implicit_pointer` support in `llvm` has been stalled out for 3 years](https://reviews.llvm.org/D84115) at this point. Do we have a better interim solution?

**Comment by jmorse** [comment]
No real other solution is available -- implicit pointers describe the fact that a source-language-construct is optimised away, trying to present these variable locations via another mechanism inevitably involves some abuse, or otherwise misrepresenting things to the user.

Alas, implicit pointers is sitting in a zone of "too hard to be low hanging fruit" and "not profitable enough given other major restructurings", hence I/colleagues haven't turned attention to it yet.

@SLTozer is this something that refined-Og and/or extended-lifetimes might affect?

**Comment by EricRahm** [comment]
@jmorse Is reviving the stalled out patchset an option? I couldn't tell if that was "basically done but with some nits" or if there were fundamental issues that needed to be ironed out.

**Comment by frobtech** [comment]
The intent of the `DW_OP_implicit_pointer` specification is that there is no necessary semantics relationship with the referenced DIE.  It's perfectly valid to always make it point to a fresh DIE that is either `DW_TAG_variable` with `DW_AT_artificial` or is just `DW_TAG_dwarf_procedure`, a DIE that exists solely to be referenced by that `DW_OP_implicit_pointer` expression.  The reason it's specified to indirect via a DIE is that it's inordinately common for there to be some otherwise-required DIE already present in the CU (such as for an actual variable in the caller of an inlined function), and thus overall DWARF size is usually optimized by referring to an existing DIE rather than creating a fresh one (or having the pointed-to value directly in the expression, which was my very first version of the `implicit_pointer` proposal when we were first defining the GNU extension that later got standardized in DWARF 5).

**Comment by dwblaikie** [comment]
Do you happen to have data on how often, in the GCC implementation, it ended up pointing to a named variable V not? (I'd have thought with many heap allocated objects being pointed to, and many references referring to names outside the current scope (such as function parameters of reference type, members of reference type for short lived objects referring back to some broader state, etc) that it'd be more common to refer to something that isn't readily referenceable as a named variable in the current scope)

In any case, yeah, I'd prefer the IR representation be more "having the pointed-to value directly in the expression" & synthesize a variable to match DWARF in the backend.

But data would be interesting to see for sure.

**Comment by SLTozer** [comment]
> @SLTozer is this something that refined-Og and/or extended-lifetimes might affect?

Yes, when the Og work has landed I believe this should be "resolved", in that Og will enable -fextend-lifetimes which should prevent `this` from being optimized out.
